test/sshkit/scp_functional_test.exs

```diff
@@ -5,33 +5,167 @@ defmodule SSHKit.SCPFunctionalTest do
+      assert :ok = SCP.download(conn, remote, local, recursive: true)
+      assert verify_transfer(conn, local, remote)
+    end
+
+    @tag boot: 1
+    test "preserve: true", %{hosts: [host]} do
+      options = [port: host.port, user: host.user, password: host.password]
+      remote = "/fixtures/file.txt"
+      local = create_random_path()
+      on_exit fn -> File.rm(local) end
+
+      {:ok, conn} = SSH.connect(host.ip, Keyword.merge(@defaults, options))
+
+      assert :ok = SCP.download(conn, remote, local, preserve: true)
+      assert verify_mode(conn, local, remote)
+      # assert verify_atime(conn, local, remote)
+      assert verify_mtime(conn, local, remote)
+    end
+
+    @tag boot: 1
+    test "recursive: true, preserve: true", %{hosts: [host]} do
+      options = [port: host.port, user: host.user, password: host.password]
+      remote = "/fixtures"
+      local = create_random_path()
+      on_exit fn -> File.rm_rf(local) end
+
+      {:ok, conn} = SSH.connect(host.ip, Keyword.merge(@defaults, options))
+
+      assert :ok = SCP.download(conn, remote, local, recursive: true, preserve: true)
+      assert verify_mode(conn, local, remote)
+      # assert verify_atime(conn, local, remote)
+      assert verify_mtime(conn, local, remote)
+    end
+  end
+
+  defp verify_transfer(conn, local, remote) do
+    command = "find <%= path %> -type f -exec shasum -a 1 {} \\; | sort | awk '{print $1}' | xargs"
+    compare_command_output(conn,
+      EEx.eval_string(command, [path: local]),
+      EEx.eval_string(command, [path: remote])
+      )
+  end
+
+  defp verify_mode(conn, local, remote) do
+    command = "find <%= path %> -type f -exec perl -e 'print join(\"\",+(stat $ARGV[0])[2,7]),\"\\n\"' {} \\; | sort | xargs"
+    compare_command_output(conn,
+      EEx.eval_string(command, [path: local]),
+      EEx.eval_string(command, [path: remote])
+      )
+  end
+
+  defp verify_atime(conn, local, remote) do
+    command = "find <%= path %> -type f -exec <%= stat %> {} \\; | cut -f1,2 | sort | xargs"
+    compare_command_output(conn,
+      EEx.eval_string(command, [path: local, stat: get_local_stat_cmd()]),
+      EEx.eval_string(command, [path: remote, stat: "stat -c '%s\t%X\t%Y'"])
+      )
+  end
+
+  defp verify_mtime(conn, local, remote) do
+    command = "find <%= path %> -type f -exec <%= stat %> {} \\; | cut -f1,3 | sort | xargs"
+    compare_command_output(conn,
+      EEx.eval_string(command, [path: local, stat: get_local_stat_cmd()]),
+      EEx.eval_string(command, [path: remote, stat: "stat -c '%s\t%X\t%Y'"])
+      )
+  end
+
+  defp compare_command_output(conn, local, remote) do
+    local_output = local |> String.to_char_list |> :os.cmd |>  to_string
+    {:ok, [stdout: remote_output], 0} = SSH.run(conn, remote)
+    assert local_output == remote_output
+
+  end
+
+  defp get_local_stat_cmd do
+    case :os.type() do
+      {:unix, :darwin} -> "stat -f '%z\t%a\t%m'"
+      _ -> "stat -c '%s\t%X\t%Y'"
+    end
+  end
```

Comment:
As for the issue of getting different timestamps on OS X and Alpine:

I've run a quick sanity check by mounting the project directory as a volume:

```shell
docker run --rm -it --volume $(pwd):/src sshkit-test-sshd /bin/sh
```

Then got the timestamps on both OS X and the shell in the container:

```shell
# OS X
❯ stat -f '%a %m' README.md 
1495494246 1495494246

# Alpine container
❯ stat -c '%X %Y' /src/README.md
1495494246 1495494246
```

Since they're exactly the same, it makes me think the setting of timestamps in our SCP up-/download implementation may not have the exact same result on two different machines.

Possible origin: [`:file.change_time/3`](http://erlang.org/doc/man/file.html#change_time-3) used in [`SCP.Download.touch!`](https://github.com/bitcrowd/sshkit.ex/blob/f09510f25af0c9e1a12b31a9808f2163520c452d/lib/sshkit/scp/download.ex#L194)

According to the Erlang docs, `:file.change_time` uses [`:file.write_file_info/2`](http://erlang.org/doc/man/file.html#write_file_info-2) under the hood, which uses `{time, local}` as the default for interpreting the Erlang datetime records that we're converting to (e.g. `{{2017, 5, 22}, {23, 4, 6}}`). So while [converting from epoch seconds to datetime record](https://github.com/bitcrowd/sshkit.ex/blob/master/lib/sshkit/scp/download.ex#L192-L193) and then using this in `:file.write_file_info`, the timezone may come into play.

Potential solution:

Use [`:file.write_file_info/3`](http://erlang.org/doc/man/file.html#write_file_info-3) with the `time: :posix` option instead of `:file.change_time` to set the `atime` and `mtime` of the file directly, without converting to a datetime object via `:calendar`.